packages/sources/finage/src/index.ts

```diff
@@ -1,10 +1,84 @@
-import { expose } from '@chainlink/ea-bootstrap'
-import { makeExecute, makeWSHandler } from './adapter'
-import * as endpoints from './endpoint'
-import { makeConfig, NAME } from './config'
-import rateLimit from './config/limits.json'
+import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
+import { IncludesFile } from '@chainlink/external-adapter-framework/adapter'
+import { PriceAdapter } from '@chainlink/external-adapter-framework/adapter'
+import { config } from './config'
+import { commodities, crypto, eod, etf, forex, stock, ukEtf } from './endpoint'
+import { PriceEndpointInputParametersDefinition } from '@chainlink/external-adapter-framework/adapter/price'
+import { AdapterParams } from '@chainlink/external-adapter-framework/adapter/types'
+import {
+  AdapterRequest,
+  AdapterRequestContext,
+  AdapterResponse,
+} from '@chainlink/external-adapter-framework/util'
+import { CustomSettingsDefinition } from '@chainlink/external-adapter-framework/config'
 
-const adapterContext = { name: NAME, rateLimit }
+export type PriceAdapterRequest<T> = AdapterRequest<T> & {
+  requestContext: AdapterRequestContext<T> & {
+    priceMeta: {
+      inverse: boolean
+    }
+  }
+}
 
-const { server } = expose(adapterContext, makeExecute(), makeWSHandler())
-export { NAME, makeExecute, makeConfig, server, endpoints }
+class FinageAdapter<T extends CustomSettingsDefinition> extends PriceAdapter<T> {
+  constructor(
+    params: AdapterParams<T> & {
+      includes?: IncludesFile
+    },
+  ) {
+    super(params)
+
+    if (params.includes) {
+      throw Error(
+        'Includes.json cannot be used for inverse behavior due to alternate implementation. See endpoint/forex.ts',
+      )
+    }
+  }
+
+  override async handleRequest(
+    req: PriceAdapterRequest<PriceEndpointInputParametersDefinition>,
+    replySent: Promise<unknown>,
+  ): Promise<AdapterResponse> {
+    /* 
+    overrides PriceAdapter.handleRequest [https://github.com/smartcontractkit/ea-framework-js/blob/main/src/adapter/price.ts#L160] 
+    to implement inverse behavior for forex endpoint see: endpoint/forex.ts/excludesMap
```

Comment:
Is the reason we're having to override this because we have an excludes map, rather than an includes map?

Reply:
I believe so. I think this was added by Subarna when aligning v2 and v3. The `handleRequest` method requires an `includesMap` to be present for inverses which Finage doesn't seem to have so I think this logic was carried over from the framework just without the `includesMap` check.

Reply:
This could potentially lead to some confusion though if an includes map is added in the future though. Then the responses may get inverted twice for `forex`. The cleaner way code-wise but harder to maintain would be to use the includes map as intended and populate it with the forex pairs we currently have jobs for that need to be inverted. Maybe generating an includes map for all or most supported forex pairs on Finage would help with maintenance too.

Reply:
But think that should be a separate effort from the upgrade so I'll mark this comment as resolved